nakadi-java-client/src/main/java/nakadi/ExponentialRetry.java

```diff
@@ -45,27 +43,33 @@ public long maxIntervalMillis() {
     return maxInterval;
   }
 
+  long workingTime() {
+    return lastBackoff - startTime;
+  }
+
   public boolean isFinished() {
-    return workingAttempts >= maxAttempts || workingTime >= maxTime;
+    return workingAttempts >= maxAttempts || workingTime() >= maxTime;
   }
 
   public long nextBackoffMillis() {
+    return nextBackOffMillis(System.currentTimeMillis());
+  }
 
+  long nextBackOffMillis(long nowMillis) {
     if (startTime == 0L) {
-      startTime = System.currentTimeMillis();
-    } else {
-      workingTime += (System.currentTimeMillis() - startTime);
+      startTime = nowMillis;
     }
+    lastBackoff = nowMillis;
 
     if (isFinished()) {
       return STOP;
     }
 
-    workingInterval = unit.toMillis(workingInterval) * (workingAttempts * workingAttempts);
+    workingInterval = workingInterval * (workingAttempts * workingAttempts);
```

Comment:
This looks like too aggressive increase. `workingInterval = startingInterval * workingInterval^2` would be OK (and easier to review).